File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/arch_units/c/au_345_c_bor_i32_u64.c
/*
 * arch-unit 345  --  c  `a | b`  lhs=int32_t rhs=uint64_t
 * symbol op_356   outcome LIFTED   2 arch-opcodes
 *
 * the arch-unit, arch-opcode by arch-opcode:
 *   c.or a0, a1                          integer    operator:|
 *   c.jr ra                              integer    return
 *
 * answer: uint64_t, 64 bits.  parameters are operand bit patterns.
 */
#include "arch_units_int.h"

uint64_t au_345_c_bor_i32_u64(uint64_t p0, uint64_t p1)
{
    /* a0: operand `a` (int32_t) sign-extended to XLEN, as the ABI presents it */
    const uint64_t v1 = au_sext32(p0);
    /* a1: operand `b` (uint64_t) arrives in a1 */
    const uint64_t v2 = p1;
    const uint64_t v3 = au_or(v1, v2);
    /* the answer is uint64_t, 64 bits */
    return v3;
}

File: PseudoCoupHQ/Research/oracle/riscv/softfloat_slices/arch_units/ruby/au_345_c_bor_i32_u64.rb
# arch-unit 345  --  c  `a | b`  lhs=int32_t rhs=uint64_t
# symbol op_356   outcome LIFTED   2 arch-opcodes
#
# the arch-unit, arch-opcode by arch-opcode:
#   c.or a0, a1                          integer    operator:|
#   c.jr ra                              integer    return
#
# answer: uint64_t, 64 bits.  parameters are operand bit patterns.
require_relative 'au_int'

def au_345_c_bor_i32_u64(p0, p1)
  # a0: operand `a` (int32_t) sign-extended to XLEN, as the ABI presents it
  v1 = au_sext32(p0)
  # a1: operand `b` (uint64_t) arrives in a1
  v2 = p1
  v3 = au_or(v1, v2)
  # the answer is uint64_t, 64 bits
  v3
end
